pkg/controller/pxc/upgrade.go

```diff
@@ -115,7 +128,159 @@ func (r *ReconcilePerconaXtraDBCluster) updatePod(sfs api.StatefulApp, podSpec *
 	currentSet.Spec.Template.Spec.Affinity = pxc.PodAffinity(podSpec.Affinity, sfs)
 	currentSet.Spec.Template.Spec.Volumes = sfsVolume.Volumes
 
-	return r.client.Update(context.TODO(), currentSet)
+	err = r.client.Update(context.TODO(), currentSet)
+	if err != nil {
+		return fmt.Errorf("update error: %v", err)
+	}
+
+	if !cr.Spec.SmartUpdateEnabled() {
+		return nil
+	}
+
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: currentSet.Name, Namespace: currentSet.Namespace}, currentSet)
+	if err != nil {
+		return fmt.Errorf("failed to get sate: %v", err)
+	}
+
+	if currentSet.Status.UpdatedReplicas >= currentSet.Status.Replicas {
+		return nil
+	}
+
+	log.Info("statefullSet was changed, run smart update")
+
+	return r.smartUpdate(sfs, cr)
+}
+
+func (r *ReconcilePerconaXtraDBCluster) smartUpdate(sfs api.StatefulApp, cr *api.PerconaXtraDBCluster) error {
+	if !isPXC(sfs) {
+		return nil
+	}
+
+	primary, err := r.getPrimaryPod(cr)
+	if err != nil {
+		return fmt.Errorf("get primary pod: %v", err)
+	}
+
+	log.Info(fmt.Sprintf("primary pod is %s", primary))
+
+	list := corev1.PodList{}
+	err = r.client.List(context.TODO(),
+		&client.ListOptions{
+			Namespace:     sfs.StatefulSet().Namespace,
+			LabelSelector: labels.SelectorFromSet(sfs.Labels()),
+		},
+		&list,
+	)
+	if err != nil {
+		return fmt.Errorf("get pod list: %v", err)
+	}
+
+	waitLimit := 120
+	if cr.Spec.PXC.LivenessInitialDelaySeconds != nil {
+		waitLimit = int(*cr.Spec.PXC.LivenessInitialDelaySeconds)
+	}
+
+	var primaryPod *corev1.Pod = nil
+	for _, pod := range list.Items {
+		pod := pod
+		if strings.HasPrefix(primary, fmt.Sprintf("%s.%s.%s", pod.Name, sfs.StatefulSet().Name, sfs.StatefulSet().Namespace)) {
+			primaryPod = &pod
+		} else {
+			log.Info(fmt.Sprintf("delete secondary pod %s", pod.Name))
+			err := r.client.Delete(context.TODO(), &pod)
+			if err != nil {
+				return fmt.Errorf("delete pod: %v", err)
+			}
+
+			err = r.waitPodRestart(&pod, waitLimit)
+			if err != nil {
+				return fmt.Errorf("wait pod: %v", err)
+			}
+		}
+	}
+
+	log.Info(fmt.Sprintf("delete primary pod %s", primaryPod.Name))
+	err = r.client.Delete(context.TODO(), primaryPod)
+	if err != nil {
+		return fmt.Errorf("delete primary pod: %v", err)
+	}
+
+	err = r.waitPodRestart(primaryPod, waitLimit)
+	if err != nil {
+		return fmt.Errorf("wait pod: %v", err)
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) getPrimaryPod(cr *api.PerconaXtraDBCluster) (string, error) {
+	user := "proxyadmin"
+	host := fmt.Sprintf("%s-proxysql-unready.%s", cr.ObjectMeta.Name, cr.Namespace)
+
+	var dbatabase queries.Database
+	i := 0
+	for {
+		db, err := queries.New(r.client, cr.Namespace, cr.Spec.SecretsName, user, host, 6032)
+		if err != nil && i < int(cr.Spec.ProxySQL.Size) {
+			time.Sleep(time.Second)
+		} else if err != nil && i == int(cr.Spec.ProxySQL.Size) {
+			return "", err
+		} else {
+			dbatabase = db
+			break
+		}
+
+		i++
+	}
```

Comment:
Because we access proxysql-unready service there can be situation when service return us the IP if failed/unready/etc proxysql pod. So we should retry `cr.Spec.ProxySQL.Size` times.